src/Node.jl

```diff
@@ -63,11 +68,42 @@ to your type.
 - `leaf_hash` and `branch_hash`
 - `preserve_sharing`
 """
-abstract type AbstractExpressionNode{T} <: AbstractNode end
+abstract type AbstractExpressionNode{T,D} <: AbstractNode{D} end
+
+mutable struct Node{T,D} <: AbstractExpressionNode{T,D}
+    degree::UInt8  # 0 for constant/variable, 1 for cos/sin, 2 for +/* etc.
+    constant::Bool  # false if variable
+    val::T  # If is a constant, this stores the actual value
+    feature::UInt16  # (Possibly undefined) If is a variable (e.g., x in cos(x)), this stores the feature index.
+    op::UInt8  # (Possibly undefined) If operator, this is the index of the operator in the degree-specific operator enum
+    children::NTuple{D,Base.RefValue{Node{T,D}}}  # Children nodes
+
+    #################
+    ## Constructors:
+    #################
+    #Node{_T,_D}() where {_T,_D} = new{_T,_D::Int}()
+    Node{_T,_D}() where {_T,_D} = (x = new{_T,_D::Int}(); x.children = ntuple(i -> Ref{Node{_T,_D}}(), Val(max_degree(Node))); x)
+end
+
+mutable struct GraphNode{T,D} <: AbstractExpressionNode{T,D}
+    degree::UInt8  # 0 for constant/variable, 1 for cos/sin, 2 for +/* etc.
+    constant::Bool  # false if variable
+    val::T  # If is a constant, this stores the actual value
+    feature::UInt16  # (Possibly undefined) If is a variable (e.g., x in cos(x)), this stores the feature index.
+    op::UInt8  # (Possibly undefined) If operator, this is the index of the operator in the degree-specific operator enum
+    children::NTuple{D,Base.RefValue{GraphNode{T,D}}}  # Children nodes
+    visited::Bool  # search accounting, initialised to false
+    cache::AbstractArray{T}
```

Comment:
Since this might not be backwards compatible, I wonder if an alternative would be to convert to a val type (`T`) where instead of `val::T`, you have `val::Tuple{T,Bool,Array{T,1}}`? Then you wouldn’t need to modify the `GraphNode` itself.